executor/analyze.go

```diff
@@ -496,22 +527,465 @@ type AnalyzeFastTask struct {
 	wg              *sync.WaitGroup
 	sampLocs        chan *tikv.KeyLocation
-	sampLocRowCount uint64
-	tasks           chan *AnalyzeFastTask
+	rowCount        uint64
+	sampCursor      int32
+	sampTasks       []*AnalyzeFastTask
 	scanTasks       []*tikv.KeyLocation
+	collectors      []*statistics.SampleCollector
+}
+
+func (e *AnalyzeFastExec) getSampRegionsRowCount(bo *tikv.Backoffer, needRebuild *bool, err *error, sampTasks *[]*AnalyzeFastTask) {
+	defer func() {
+		e.wg.Done()
+		if *needRebuild == true {
+			for ok := true; ok; _, ok = <-e.sampLocs {
+				// Do nothing, just clear the channel.
+			}
+		}
+	}()
+	client := e.ctx.GetStore().(tikv.Storage).GetTiKVClient()
+	for {
+		loc, ok := <-e.sampLocs
+		if !ok {
+			return
+		}
+		req := &tikvrpc.Request{
+			Type: tikvrpc.CmdDebugGetRegionProperties,
+			DebugGetRegionProperties: &debugpb.GetRegionPropertiesRequest{
+				RegionId: loc.Region.GetID(),
+			},
+		}
+		var resp *tikvrpc.Response
+		var rpcCtx *tikv.RPCContext
+		rpcCtx, *err = e.cache.GetRPCContext(bo, loc.Region)
+		if *err != nil {
+			return
+		}
+		ctx := context.Background()
+		resp, *err = client.SendRequest(ctx, rpcCtx.Addr, req, tikv.ReadTimeoutMedium)
+		if *err != nil {
+			return
+		}
+		if resp.DebugGetRegionProperties == nil || len(resp.DebugGetRegionProperties.Props) == 0 {
+			*needRebuild = true
+			return
+		}
+		for _, prop := range resp.DebugGetRegionProperties.Props {
+			if prop.Name == "mvcc.num_rows" || prop.Name == "num_rows" {
+				var cnt uint64
+				cnt, *err = strconv.ParseUint(prop.Value, 10, 64)
+				if *err != nil {
+					return
+				}
+				newCount := atomic.AddUint64(&e.rowCount, cnt)
+				task := &AnalyzeFastTask{
+					Location:  loc,
+					LRowCount: newCount - cnt,
+					RRowCount: newCount,
+				}
+				*sampTasks = append(*sampTasks, task)
+			}
+		}
+	}
+}
+
+// buildSampTask return tow variable, the first bool is whether the task meeting region error
+// and need to rebuild.
+func (e *AnalyzeFastExec) buildSampTask() (bool, error) {
+	// Do get regions row count.
+	bo := tikv.NewBackoffer(context.Background(), 500)
+	atomic.StoreUint64(&e.rowCount, 0)
+	needRebuildForRoutine := make([]bool, e.concurrency)
+	errs := make([]error, e.concurrency)
+	sampTasksForRoutine := make([][]*AnalyzeFastTask, e.concurrency)
+	e.sampLocs = make(chan *tikv.KeyLocation, e.concurrency)
+	e.wg.Add(e.concurrency)
+	for i := 0; i < e.concurrency; i++ {
+		go e.getSampRegionsRowCount(bo, &needRebuildForRoutine[i], &errs[i], &sampTasksForRoutine[i])
+	}
+
+	store, _ := e.ctx.GetStore().(tikv.Storage)
+	e.cache = store.GetRegionCache()
+	startKey, endKey := tablecodec.GetTableHandleKeyRange(e.physicalTableID)
+	// extract all regions contain the table
+	e.scanTasks = e.scanTasks[:0]
+	e.sampTasks = e.sampTasks[:0]
+	targetKey := startKey
+	for {
+		// Search for the region which contains the targetKey.
+		loc, err := e.cache.LocateKey(bo, targetKey)
+		if err != nil {
+			return false, err
```

Comment:
before returning this error, we should close `e.sampLoks` to avoid goroutine `e. getSampRegionsRowCount()` leak

Reply:
Now it extracts all these logic to `defer` function.